bubo-ht: use bitset instead of linked list

davidvgalbraith opened the pull request:
this decreases memory usage by ~20%

@demmer

Commit:
bubo-ht: use bitset instead of linked list

this decreases memory usage by ~20%

demmer commented on src/bubo-ht.h:
```diff
@@ -166,185 +114,100 @@ class BuboHashSet {
 
 protected:
     struct Entry {
         const BYTE* val_;
```
There's no need to have an Entry struct any more -- The table should just be an array of `BYTE*`

demmer commented on src/bubo-ht.h:
```diff
@@ -166,185 +114,100 @@ class BuboHashSet {
     Entry* table_;
 
     BlobStore* blob_store_;
+    boost::dynamic_bitset<> bit_set_;
```
I'd find it more readable if this were called `occupied_` or `used_`.

demmer commented on src/bubo-ht.h:
```diff
@@ -166,185 +114,100 @@ class BuboHashSet {
-            // not found, but spine has an entry. insert new item right next to spine.
-            table[index].next_ = new Entry(value, table[index].next_);
-            table_collisions_ ++;
+        while (bit_set_[index]) {
```
This will infinitely loop if there are already `max_table_size_` entries in the table, i.e. all slots are full and you can't resize any more, so you need to handle that case. Also adding a test  of course.

demmer replied:
Oh I see -- you guard against this in `maybe_resize`. I guess that's ok.

demmer replied:
OK but one more thing -- suppose you previously inserted an entry at `index`, then deleted it, then added another entry. Your slot will be empty, but the bit will still be set.

I think to solve this you'll want to change the above loop to `while (table_[index] != NULL && bit_set_[index == true)`.

demmer commented on src/bubo-ht.h:
```diff
@@ -111,53 +79,33 @@ class BuboHashSet {
-                erase_next(erase_entry);
-            }
-            num_entries_ --;
+            Entry* entry_to_erase = &table_[index];
```
This isn't going to work if the entry to erase is in a slot after `index` due to a collision since you'll erase the entry at `index`, not the one you want.

Commit:
get rid of Entry

Commit:
bit_set_ -> occupied_

Commit:
Fix delete
remove the entry where the entry to delete is, not
where we started looking for it

davidvgalbraith commented:
Ok, fixed it up.

davidvgalbraith commented:
It turned out there had been some software engineering applied here, so it was trivial to make a Bubo that always collides and test it.
